Sources/App/routes.swift

```diff
@@ -34,102 +34,107 @@ func routes(_ app: Application) throws {
-        req.session.data["username"] = username
-        req.session.data["userID"] = userID.uuidString
-        return MakeCredentialResponse(userID: userID.uuidString.base64String(), challenge: challenge)
+
+        let user = User(username: username)
+        try await user.save(on: req.db)
+
+        let (options, sessionData) = try req.webAuthn.beginRegistration(user: user)
+
+        req.logger.debug("Challenge is \(options.challenge)")
+
+        req.session.data["challenge"] = sessionData.challenge
+        req.session.data["userID"] = sessionData.userID
+
+        return options
     }
-    
+
     // step 2 for registration
     authSessionRoutes.post("makeCredential") { req -> HTTPStatus in
         guard let challenge = req.session.data["challenge"] else {
             throw Abort(.unauthorized)
         }
-        let registerData = try req.content.decode(RegisterWebAuthnCredentialData.self)
-        
+        let registerData = try req.content.decode(RegistrationResponse.self)
+
         guard let origin = Environment.get("ORIGIN") else {
             throw Abort(.internalServerError)
         }
-        
-        let credential = try WebAuthn.parseRegisterCredentials(registerData, challengeProvided: challenge, origin: origin, logger: req.logger)
-        
-        guard let username = req.session.data["username"], let userIDString = req.session.data["userID"], let userID = UUID(uuidString: userIDString) else {
+
+        let credential = try req.webAuthn.parseRegisterCredentials(registerData, challengeProvided: challenge, origin: origin, logger: req.logger)
+
+        guard let userIDString = req.session.data["userID"],
+            let userID = UUID(uuidString: userIDString),
+            let user = try await User.find(userID, on: req.db) else {
             throw Abort(.badRequest)
         }
-        
-        let user = User(id: userID, username: username)
-        try await user.save(on: req.db)
-        
+
         let webAuthnCredential = WebAuthnCredential(id: credential.credentialID, publicKey: credential.publicKey.pemRepresentation, userID: userID)
         try await webAuthnCredential.save(on: req.db)
-        
+
         req.auth.login(user)
-        
+
         return .ok
     }
-    
+
     // step 1 for authentication
     authSessionRoutes.get("authenticate") { req -> StartAuthenticateResponse in
         let username = try req.query.get(String.self, at: "username")
         guard let user = try await User.query(on: req.db).filter(\.$username == username).first() else {
             throw Abort(.unauthorized)
         }
-        let challenge = [UInt8].random(count: 32).base64
-        let encodedChallenge = challenge.replacingOccurrences(of: "+", with: "-")
-            .replacingOccurrences(of: "/", with: "_")
-            .replacingOccurrences(of: "=", with: "")
+        let challenge = try req.webAuthn.generateChallenge()
+        let encodedChallenge = challenge.base64URLEncodedString()
```

Comment:
This should probably just be `req.webAuthn.generateChallengeString()`

Reply:
👍🏼